mypy.ini

```diff
@@ -0,0 +1,1070 @@
+[mypy]
+python_version = 3.9
+pretty = false
+exclude = (?x)(
+          docs|
+          examples
+          )
+
+enable_error_code = ignore-without-code
+
+# We can get to passing mypy's strict mode by incrementally changing the
+# following global settings to true. This comes from mypy's documentation.
+
+warn_unused_configs = true
+warn_redundant_casts = true
+warn_unused_ignores = true
+strict_equality = true
+check_untyped_defs = false
+disallow_subclassing_any = false
+disallow_untyped_decorators = false
+disallow_any_generics = false
+disallow_untyped_calls = false
+disallow_incomplete_defs = false
+disallow_untyped_defs = false
+no_implicit_reexport = false
+warn_return_any = false
+extra_checks = false
```

Comment:
These settings come from [mypy's doc page on introducing enabling mypy on an existing code base]. The [doc page on mypy configuration files](https://mypy.readthedocs.io/en/stable/config_file.html) describes the individual settings.

Reply:
> My one larger suggestion is to have the configurations set on a per-subpackage basis rather than globally for all of Astropy.

@nstarman — just to make sure I understand, did you mean having a block of settings like this so that configuration variables like `warn_redundant_casts`, `strict_equality`, `check_untyped_defs`, etc. could be defined on a per-subpackage basis?  I like that idea, and I think I'll probably apply that to PlasmaPy as well.  

Alternatively, did you mean to have the per-file ignores become per-subpackage ignores like in #12971?  That would also be a valid approach.  The ignores would become less fine-grained, but `mypy.ini` would become shorter and probably more maintainable.

Thank you!

Reply:
The former 👍. I just imagine that some subpackages will achieve `strict=True` well before others.